Make "switch costume" and "switch backdrop" blocks compatible with 2.0

joker314 opened the pull request:
### Resolves

Resolves #1490 

### Proposed Changes

- `switch backdrop to` block is (EDIT: [almost](https://github.com/LLK/scratch-vm/pull/1517#discussion_r223779175)) compatible with 2.0 behaviour
- `switch costume to` block is compatible with 2.0 behaviour
- unit tests guard against future incompatibility
- make "random backdrop" have lower priority than a costume with that name

See [2.0 behaviour description](https://github.com/LLK/scratch-vm/issues/1490#issuecomment-413679793).

### Reason for Changes

Making sure that 2.0 projects still work in 3.0.

I placed the "random backdrop" feature as _less_ important than a costume name. This means that if a 2.0 project already had a backdrop with the name "random backdrop", references to it by name will still hold. Was this a good idea?

This is different to how "previous backdrop" and "next backdrop" behave -- they can't be overridden (per 2.0); however this is the same behaviour as "previous costume" and "next costume", which _can._

### Test Coverage

- 2 automatic unit tests were added. They all pass after my changes. Expand below to see which failed before (more than was initially documented in the issue).
 
<details><summary>Assertions which failed prior to changes</summary>
<h4>Costumes</h4>
<ul>
    <li> <code>switch costume to NaN</code> (e.g. <code>0/0</code>) should flawlessly switch to the first costume. </li>
    <li> <code>switch costume to Infinity</code> (e.g. <code>1/0</code>) should flawlessly switch to the first costume. </li>
    <li> <code>switch costume to -Infinity</code> (e.g. <code>-1/0</code>) should flawlessly switch to the first costume. </li>
    <li> <code>switch <strong>costume</strong> to previous <strong>backdrop</strong></code> </li> should do nothing
    <li> <code>switch <strong>costume</strong> to next <strong>backdrop</strong></code> should do nothing</li>
    <li> Strings should only be interpreted as indices if they contain at least one numeric digit.</li>
    <li> <code>switch costume to &lt;1 = 0></code> should switch to the costume named "false" if it exists #1490 </li></ul>
  <h4>Backdrops</h4>
    <ul><li> <s><code>switch backdrop to [previous backdrop]</code> decrements the backdrop <em>even if there's a backdrop with that name</em> </s></li>
    <li> <s><code>switch backdrop to [next backdrop]</code> increments the backdrop <em>even if there's a backdrop with that name</em></s></li>
    <li> <code>switch backdrop to NaN</code> (e.g. <code>0/0</code>) should flawlessly switch to the first backdrop.</li>
    <li> <code>switch backdrop to Infinity</code> (e.g. <code>1/0</code>) should flawlessly switch to the first backdrop.</li>
    <li> <code>switch backdrop to -Infinity</code> (e.g. <code>-1/0</code>) should flawlessly switch to the first backdrop.</li></ul>
</details>
<hr>

- Manually tested with 239254767
- Manually tested with 239255071
- Unable to manually test with 86981455 as it wasn't a minimal example
- Manually confirmed that switching costume by name still worked

joker314 committed:
Create unit tests for SWITCH COSTUME and SWITCH BACKDROP

joker314 committed:
Make SWITCH COSTUME and SWITCH BACKDROP compatible with 2.0

joker314 commented:
Regarding the unit tests added, would it be a good idea to split them into a separate file -- and, since the backdrop and costume tests are quite similar, use a loop/function for the common bits?

kchadha commented on src/blocks/scratch3_looks.js:
```diff
@@ -344,73 +344,94 @@ class Scratch3LooksBlocks {
-                const numCostumes = target.getCostumes().length;
+            } else if (requestedCostume === 'previous costume') {
+                target.setCostume(target.currentCostume - 1);
+            } else if (!isNaN(requestedCostume) &&
```
The first part of this if will always be true because `NaN` is of type `'number'` which you already checked for in the parent if.

joker314 replied:
`isNaN` casts its argument to a number before checking if it is NaN. Hence, `isNaN("meow") -> true`. Many complain that the function is named improperly and hence its behaviour is unintuitive.

`Number.isNaN` does not do this casting. Hence, this PR could do the casting explicitly, and then use `Number.isNaN`.

Note: `Number.isNaN` is not supported in IE11. Is this okay?

kchadha replied:
Oh I didn't know about `isNaN` implicitly casting! Thanks. =] In that case, do you mind adding a comment in the places where you use it like this?

kchadha commented on src/blocks/scratch3_looks.js:
```diff
@@ -344,73 +344,94 @@ class Scratch3LooksBlocks {
+            } else if (requestedCostume === 'previous costume') {
+                target.setCostume(target.currentCostume - 1);
+            } else if (!isNaN(requestedCostume) &&
+                        (typeof requestedCostume !== 'string' || /\d/g.test(requestedCostume))) {
```
I'm not sure why you need the regex test here.

If you're tying to see if it's a string representation of a number e.g. "12", you could try casting requestedCostume to a number and see if it results in `NaN` or not.

kchadha replied:
e.g. in the [old code here](https://github.com/LLK/scratch-vm/pull/1517/files?utf8=%E2%9C%93&diff=split#diff-8d62e377c35d800fc8cfe63d4e98232fL376)

joker314 replied:
Scratch thinks that empty space is not a number (whereas JavaScript thinks it's 0). Looking for a digit was [the algorithm used by 2.0](https://github.com/LLK/scratch-flash/blob/develop/src/interpreter/Interpreter.as#L399-L413) -- but it didn't use a regular expression in its implementation.

I think this regular expression should be moved to `util/cast.js`

kchadha replied:
Hmmm it seems like maybe we could use a combination of `parseInt` and `isNaN`. Though, unfortunately `parseInt` does not work the way we want for boolean inputs, so we would have to deal with those explicitly.

I'm also okay with leaving this code as is now that I know the reason.

In either case, could you add a comment here explaining the digit check (e.g. the fact that whitespace is treated as 0 when casting to a number, and we don't want that)

joker314 replied:
This is a good point. I think I'll use the `Cast` utility for this -- `Cast.isWhiteSpace`.

joker314 replied:
Since I'm now using the very readable utility function, I decided to remove the explanation that JS casts whitespace to 0.

kchadha commented on src/blocks/scratch3_looks.js:
```diff
@@ -344,73 +344,94 @@ class Scratch3LooksBlocks {
+        } else if (requestedBackdrop === 'next backdrop') {
+            stage.setCostume(stage.currentCostume + 1);
+        } else if (requestedBackdrop === 'previous backdrop') {
+            stage.setCostume(stage.currentCostume - 1);
```
This behavior is consistent with Scratch 2.0, as you noted here: 

https://github.com/LLK/scratch-vm/issues/1490#issuecomment-413714669

But from a usability/consistency perspective this definitely seems like a bug.

My thought is to make backdrops more consistent with costumes in this regard (e.g. if you have a costume named 'next costume' you can switch to it).
@thisandagain do you have any thoughts on this?

At the very least, the `random backdrop` option should be handled the same way we decide to handle `previous backdrop` and `next backdrop`

thisandagain replied:
Agreed @kchadha. I think this is a place I'd be ok with a breaking change. We should document it here: https://github.com/LLK/scratch-vm/wiki/Compatibility

thisandagain commented:
Thanks a ton @joker314. In addition to the PR, I really appreciate the documentation and research in https://github.com/LLK/scratch-vm/issues/1490. 😄

kchadha commented on test/unit/blocks_looks.js:
```diff
@@ -23,6 +25,146 @@ const fakeRuntime = {
+    // Non-existant costumes do nothing
+    t.strictEqual(testCostume(['a', 'b', 'c', 'd'], 'e', 3), 3);
+
+    // Difference between string and numeric arguments
```
These tests are awesome and very thorough!

Could you clarify this comment a little bit e.g. "Number arguments are always treated as the costume index. String arguments are treated as the name of the costume, otherwise converted to the costume index if possible"  and add one more test case to illustrate all the possible cases here
`t.strictEqual(testCostume(['a', 'b', 'c'], '2'), 2);`

kchadha commented on test/unit/blocks_looks.js:
```diff
@@ -23,6 +25,146 @@ const fakeRuntime = {
+
+    // Backdrop indices should wrap around.
+    t.strictEqual(testBackdrop(['a', 'b', 'c', 'd'], -1), 3);
+    t.strictEqual(testBackdrop(['a', 'b', 'c', 'd'], -4), 4);
```
Yay for fixing the negative number wrap-around bug for backdrops!

kchadha commented on test/unit/blocks_looks.js:
```diff
@@ -23,6 +25,146 @@ const fakeRuntime = {
+     * @param {number} [currentCostume=1] The 1-indexed default costume for the sprite to start at.
+     * @return {number} The 1-indexed costume index on which the sprite lands.
+     */
+    const testCostume = (costumes, arg, currentCostume = 1) => {
```
`testCostume` and `testBackdrop` can be helpers outside of this test definition. You can also add one more optional argument at the end of testCostume to swap the relevant costume/backdrop bits so that the two functions can share code.

E.g. 
```
const testCostume = (costumes, arg, currentCostume = 1, costumeOrBackdrop = 'costume') => { ... };

const testBackdrop = (backdrop, arg, currentBackdrop = 1) => testCostume(backdrop, arg, currentBackdrop, 'backdrop');
```

I also think this code is okay as is though.

kchadha reviewed (changes requested):
@joker314 Thanks for making this pull request and for detailing the Scratch 2.0 behavior in the comment on #1490. I apologize for the delay on this review.

I am requesting some minor changes/code cleanup, and would also like to request adding more comments in the code to illustrate what kinds of cases you're trying to address in which parts of the code.

Thanks for adding the thorough unit tests, they're awesome!

I added a comment about maybe making the `switch backdrop` behavior similar to `switch costume` in how it handles backdrops named 'previous backdrop' and 'next backdrop' but I think it's also okay to file that as a follow-up separate issue since what you have now is consistent with Scratch 2.0.

joker314 committed:
test: add more comments, add additional test

joker314 committed:
refactor: use utility for whitespace-testing

joker314 committed:
refactor: move helper functions for tests

Move the helper functions `testCostume` and `testBackdrop` outside of
their test definitions. In addition, define the latter in terms of the
former.

joker314 committed:
style(blocks_looks.js): add semicolon/space

joker314 committed:
refactor: use better whitespace-test for backdrops too

joker314 committed:
feat: Allow switching to specially named backdrops

The `switch costume` block accepts special values like "next costume" and
"previous costume". If you create a costume with these names, these take
priority over the special values. However, the `switch backdrop` block
keeps these special values for values like "next backdrop", "previous
backdrop", "random backdrop". It is impossible to navigate to such a
backdrop by name via block. This commit also modifies tests to allow for
this.

BREAKING CHANGE: specially-named backdrops can now be navigated

joker314 commented:
Thank you very much for the review @kchadha. I've addressed your thoughtful feedback, and have introduced the breaking change which makes backdrop special arguments be handled in a similar way to costumes.

This pull request is ready for re-review.

Here's a base for what one might want to include in the [Compatibility wiki page](https://github.com/LLK/scratch-vm/wiki/Compatibility), if it's useful:

> ### Backdrops named 'previous backdrop' and 'next backdrop' can be switched to via blocks
>
> In Scratch 2.0, `switch backdrop to [next backdrop]` would switch to the backdrop sequentially following the current backdrop, even if there was a backdrop literally named "next backdrop". The same behaviour is seen in "previous backdrop".
>
> This differed to the behaviour of the `switch costume` block, which would not perform the block input's special function if there existed a costume with the corresponding name.
>
> To make the blocks more intuitive, the `switch backdrop` block now uses the `switch costume` behaviour. This is also likely to be what a user would expect, since the workaround for incrementing a backdrop is trivial and stable, whereas the workaround for switching to a backdrop by name would require trying every costume inefficiently until it is found (unless the order in which the backdrops have been arranged is hard-coded)

joker314 committed:
docs: Provide more/clearer comments

We ensure that all code whose purpose may be confusing to grasp is
commented; and we remove information that is no longer required.

kchadha reviewed (approved):
@joker314 Thank you for making the requested changes! This PR looks good to go.
